src/EditorFeatures/CSharp/StringCopyPaste/KnownSourcePasteProcessor.cs

```diff
@@ -0,0 +1,301 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Indentation;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.StringCopyPaste
+{
+    using static StringCopyPasteHelpers;
+    using static SyntaxFactory;
+
+    internal class KnownSourcePasteProcessor : AbstractPasteProcessor
+    {
+        private readonly ExpressionSyntax _stringExpressionCopiedFrom;
+        private readonly ITextSnapshot _snapshotCopiedFrom;
+
+        public KnownSourcePasteProcessor(
+            string newLine,
+            IndentationOptions indentationOptions,
+            ITextSnapshot2 snapshotBeforePaste,
+            ITextSnapshot2 snapshotAfterPaste,
+            Document documentBeforePaste,
+            Document documentAfterPaste,
+            ExpressionSyntax stringExpressionBeforePaste,
+            ExpressionSyntax stringExpressionCopiedFrom,
+            ITextSnapshot snapshotCopiedFrom)
+            : base(newLine, indentationOptions, snapshotBeforePaste, snapshotAfterPaste, documentBeforePaste, documentAfterPaste, stringExpressionBeforePaste)
+        {
+            _stringExpressionCopiedFrom = stringExpressionCopiedFrom;
+            _snapshotCopiedFrom = snapshotCopiedFrom;
+        }
+
+        public override ImmutableArray<TextChange> GetEdits(CancellationToken cancellationToken)
+        {
+            // For pastes into non-raw strings, we can just determine how the change should be escaped in-line at that
+            // same location the paste originally happened at.  For raw-strings things get more complex as we have to
+            // deal with things like indentation and potentially adding newlines to make things legal.
+
+            // Smart Pasting into raw string not supported yet.  
```

Comment:
this will come in the next PR.  It is complex enough that i want to make that easier to review on its own.

Reply:
There is a comedic irony in flagging things for later to be easier to review, in a PR with 55 commits and bits of code surrounded by `#if false` 😛

I guess we should be thankful though, it presumably could have been worse!

Reply:
still irony here.  still coming in next PR :)